### DISCUSSION TITLE
How to unlink with all linked models in `update`?

### DISCUSSION THREAD

**Discussion by teenjuna** [Brendonovich/prisma-client-rust#110] [Q&A] [answered]
For example, I have an user that can be linked with many posts. In the update parameters, I can unlink it from posts by providing the ids of linked posts. In order for that to happen, I need to fetch linked posts beforehand. Is there an easier solution that doesn't involve calling the database two times?

**Comment by Brendonovich** [comment, marked as answer]
Oh, I think this is actually a problem with PCR. The Prisma docs say that updateMany takes a WhereInput, but I have taking UniqueWhereInputs. Will get that fixed.

**Reply by teenjuna** [reply]
Nice 😎. The library is getting better in small steps! Should I probably create an issue from the discussion?

**Reply by Brendonovich** [reply]
Yes please! I'll get it fixed for 0.6.0.

**Reply by teenjuna** [reply]
@Brendonovich I saw a fix commit in `v0.6.0` branch. Is `unlink(vec![])` supposed to work? Because if it is, then there is another problem. I get this weird runtime error in my codebase:
```
Error executing query: P2009 - Failed to validate the query: `Unable to match input value to any allowed input type for the field. Parse errors: [Query parsing/validation error at `Mutation.updateOneProduct.data.ProductUpdateInput.categories.CategoryUpdateManyWithoutProductsInput.disconnect`: Unable to match input value to any allowed input type for the field. Parse errors: [Query parsing/validation error at `Mutation.updateOneProduct.data.ProductUpdateInput.categories.CategoryUpdateManyWithoutProductsInput.disconnect.CategoryWhereUniqueInput`: Expected exactly one field to be present, got 0., Query parsing/validation error at `Mutation.updateOneProduct.data.ProductUpdateInput.categories.CategoryUpdateManyWithoutProductsInput.disconnect`: Value types mismatch. Have: Object({}), want: Object(CategoryWhereUniqueInput)], Query parsing/validation error at `Mutation.updateOneProduct.data.ProductUncheckedUpdateInput.categories.CategoryUncheckedUpdateManyWithoutProductsInput.disconnect`: Unable to match input value to any allowed input type for the field. Parse errors: [Query parsing/validation error at `Mutation.updateOneProduct.data.ProductUncheckedUpdateInput.categories.CategoryUncheckedUpdateManyWithoutProductsInput.disconnect.CategoryWhereUniqueInput`: Expected exactly one field to be present, got 0., Query parsing/validation error at `Mutation.updateOneProduct.data.ProductUncheckedUpdateInput.categories.CategoryUncheckedUpdateManyWithoutProductsInput.disconnect`: Value types mismatch. Have: Object({}), want: Object(CategoryWhereUniqueInput)]]` at `Mutation.updateOneProduct.data
```

**Reply by Brendonovich** [reply]
It is meant to work, but it doesn't for a reason that took me by surprise until now.
At the moment I convert the `Vec` passed to `unlink` to an object and give that to Prisma since that's what fits the GraphQL schema. Turns out that the GraphQL schema is a sneaky liar and `unlink` (or `disconnect`) as it's called internally can accept an array too and that's what the JS, Python and Go clients use. Will do that instead then.

**Reply by teenjuna** [reply]
But why did you change `WhereParam` to `UniqueWhereParam` again? Now `unlink(vec![])` compiles and gives no runtime errors, but it makes no effect 😅

EDIT: I tried changing it back to `WhereParam`, but it still doesn't unlink anything. In Prisma docs, there is [`set`](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#disconnect-all-related-records) method for this. Maybe we need something similar?

Also, side question. Why did you call these methods `link` and `unlink` instead of `connect` and `disconnect` like in original client? If keeping the same naming is not a goal, I wanted to suggest renaming some comparison methods to match Rust's names in traits like `PartialOrd` and `PartialEq`, eg `equals` -> `eq`, `not` -> `ne`, `gte` -> `ge`. I felt a small tension when first tried using those with current naming.

**Reply by Brendonovich** [reply]
I changed it because it's wrong ([see here](https://github.com/Brendonovich/prisma-client-rust/issues/112#issuecomment-1191800854)). `unlink(vec![])` shouldn't give an error, but it also shouldn't do anything. As you suggested, I may add a `set(Vec<UniqueWhereParam>)` method.

I called them `link` and `unlink` as that is what they were called in the Go client, which this project was heavily based off of in the beginning. I'm open to using more of Prisma's naming for 0.6.0.

**Reply by teenjuna** [reply]
Got you. While there's no `set` method, I have another question (sorry for bothering): is `unlink` meant to work with subsequent `link` call? I'm trying to emulate `set` behaviour by first fetching the saved model, then calling `update` with updates like `vec![unlink(all_linked_models), link(models_to_link)]`, but it takes no effect even if argument to `link` is and empty vector. If I remove the `link`, it works.

**Reply by Brendonovich** [reply]
Well here's why that happens, lol
<img width="859" alt="image" src="https://user-images.githubusercontent.com/14191578/180367341-c1431696-9aa8-42d0-b5ae-3335459db924.png">

**Reply by teenjuna** [reply]
But they have different ID's, shouldn't that work?

**Reply by Brendonovich** [reply]
The problem is that the link and unlink are being treated as separate updates for the `posts` relation, rather than being grouped together as 2 updates under the 1 relation. In this case, disconnect is being discard in favour of connect.

Currently, the data is resolved to an object like this:
```
posts: {
    connect: []
},
posts: {
    disconnect: []
}
```
But it needs to look like this:
```
posts: {
    connect: [],
    disconnect: [],
}
```

**Reply by teenjuna** [reply]
Now I get it. Is that the issue in codegen?

**Reply by Brendonovich** [reply]
Not the codegen but the library, I'm fixing it now.

**Reply by Brendonovich** [reply]
Try the latest commit to 0.6.0

**Reply by teenjuna** [reply]
Yeah, it seems to work fine now!
